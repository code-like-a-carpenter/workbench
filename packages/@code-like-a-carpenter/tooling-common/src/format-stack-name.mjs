import camelCase from 'lodash.camelcase';
import upperFirst from 'lodash.upperfirst';

/**
 * @typedef {Object} FormatStackNameOptions
 * @property {boolean} ci
 * @property {string} [fullRef]
 * @property {string} projectName
 * @property {string} [sha]
 * @property {string} [ref]
 */

/**
 * @param {FormatStackNameOptions} options
 * @returns {string}
 */
export function formatStackName({
  ci,
  fullRef = '',
  projectName,
  sha = '',
  ref = '',
}) {
  const stackName = upperFirst(camelCase(projectName));
  let suffix = '';

  if (!ci) {
    return stackName;
  }

  if (sha !== '') {
    suffix = sha.slice(0, 7);
  }

  let branch = '';
  if (fullRef !== '') {
    branch = fullRef;
  } else if (ref) {
    branch = ref.split('/').slice(2).join('/');
  }

  if (branch) {
    if (branch.startsWith('dependabot')) {
      branch = branch.split('/').slice(2).join('/');
    }

    suffix = `${camelCase(branch.replace(/[/_]/g, '-').substring(0, 20))}-${suffix}`;
  }

  return `ci--${suffix ? `${stackName}-${suffix}` : stackName}`;
}
