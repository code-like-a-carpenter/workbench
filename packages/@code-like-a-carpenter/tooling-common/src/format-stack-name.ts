import camelCase from 'lodash.camelcase';
import upperFirst from 'lodash.upperfirst';

export interface FormatStackNameOptions {
  readonly ci: boolean;
  readonly fullRef?: string;
  readonly projectName: string;
  readonly sha?: string;
  readonly ref?: string;
}
export function formatStackName({
  ci,
  fullRef = '',
  projectName,
  sha = '',
  ref = '',
}: FormatStackNameOptions): string {
  const stackName = upperFirst(camelCase(projectName));
  let suffix = '';

  if (!ci) {
    return stackName;
  }

  if (sha !== '') {
    suffix = sha.slice(0, 7);
  }

  let branch = '';
  if (fullRef !== '') {
    branch = fullRef;
  } else if (ref) {
    branch = ref.split('/').slice(2).join('/');
  }

  if (branch) {
    if (branch.startsWith('dependabot')) {
      branch = branch.split('/').slice(2).join('/');
    }

    suffix = `${camelCase(branch.replace(/[/_]/g, '-').substring(0, 20))}-${suffix}`;
  }

  return `ci--${suffix ? `${stackName}-${suffix}` : stackName}`;
}
