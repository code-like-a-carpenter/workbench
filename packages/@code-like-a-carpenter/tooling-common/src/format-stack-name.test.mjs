import {formatStackName} from './format-stack-name.mjs';

const GITHUB_HEAD_REF = 'dependabot/npm_and_yarn/openapi-typescript-6.7.4';
const GITHUB_REF =
  'refs/heads/dependabot/npm_and_yarn/openapi-typescript-6.7.4';
const GITHUB_SHA = 'f7d69db466396454e27435e3dc4affde6a1f3263';
const BUILDKITE_BRANCH = 'dependabot/npm_and_yarn/openapi-typescript-6.7.4';
const BUILDKITE_COMMIT = 'f7d69db466396454e27435e3dc4affde6a1f3263';

const gitData = [
  {ci: true, sha: GITHUB_SHA},
  {ci: true, fullRef: GITHUB_HEAD_REF, sha: GITHUB_SHA},
  {ci: true, ref: GITHUB_REF, sha: GITHUB_SHA},
  {ci: true, sha: BUILDKITE_COMMIT},
  {ci: true, ref: BUILDKITE_BRANCH, sha: BUILDKITE_COMMIT},
];

const projectNames = [
  '@code-like-a-carpenter/tooling-common',
  '@code-like-a-carpenter/foundation-intermediate-representation',
  'aws-otel',
  'check-run-reporter',
];

const testData = projectNames
  .map((projectName) =>
    gitData.map((env) => {
      /** @type {[string, {ci: boolean, fullRef?: string, sha: string, ref?: string}]} */
      const ret = [projectName, env];
      return ret;
    })
  )
  .flat(1);

describe('formatStackName()', () => {
  it.each(projectNames)(
    'converts a project name to a stack name',
    (projectName) => {
      const stackName = formatStackName({ci: false, projectName});
      expect(stackName).toMatchSnapshot();
      expect(stackName.length).toBeLessThan(128);
    }
  );

  it.each(testData)(
    'converts a project name to a stack name in ci',
    (projectName, env) => {
      const stackName = formatStackName({projectName, ...env});
      expect(stackName).toMatchSnapshot();
      expect(stackName).toMatch(/^ci--/);
      expect(stackName.length).toBeLessThan(128);
    }
  );

  it.each(testData)(
    'converts a project name to a stack name in ci in prod',
    (projectName, env) => {
      const stackName = formatStackName({projectName, ...env, isProd: true});
      expect(stackName).toMatchSnapshot();
      expect(stackName.length).toBeLessThan(128);
    }
  );
});
