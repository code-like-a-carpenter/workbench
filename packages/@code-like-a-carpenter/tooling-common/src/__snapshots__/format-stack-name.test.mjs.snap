// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatStackName() converts a project name to a stack name 1`] = `"CodeLikeACarpenterToolingCommon"`;

exports[`formatStackName() converts a project name to a stack name 2`] = `"CodeLikeACarpenterFoundationIntermediateRepresentation"`;

exports[`formatStackName() converts a project name to a stack name 3`] = `"AwsOtel"`;

exports[`formatStackName() converts a project name to a stack name 4`] = `"CheckRunReporter"`;

exports[`formatStackName() converts a project name to a stack name in ci 1`] = `"ci--CodeLikeACarpenterToolingCommon-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 2`] = `"ci--CodeLikeACarpenterToolingCommon-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 3`] = `"ci--CodeLikeACarpenterToolingCommon-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 4`] = `"ci--CodeLikeACarpenterToolingCommon-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 5`] = `"ci--CodeLikeACarpenterToolingCommon-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 6`] = `"ci--CodeLikeACarpenterFoundationIntermediateRepresentation-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 7`] = `"ci--CodeLikeACarpenterFoundationIntermediateRepresentation-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 8`] = `"ci--CodeLikeACarpenterFoundationIntermediateRepresentation-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 9`] = `"ci--CodeLikeACarpenterFoundationIntermediateRepresentation-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 10`] = `"ci--CodeLikeACarpenterFoundationIntermediateRepresentation-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 11`] = `"ci--AwsOtel-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 12`] = `"ci--AwsOtel-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 13`] = `"ci--AwsOtel-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 14`] = `"ci--AwsOtel-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 15`] = `"ci--AwsOtel-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 16`] = `"ci--CheckRunReporter-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 17`] = `"ci--CheckRunReporter-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 18`] = `"ci--CheckRunReporter-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 19`] = `"ci--CheckRunReporter-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci 20`] = `"ci--CheckRunReporter-openapiTypescript6-f7d69db"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 1`] = `"CodeLikeACarpenterToolingCommon"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 2`] = `"CodeLikeACarpenterToolingCommon"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 3`] = `"CodeLikeACarpenterToolingCommon"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 4`] = `"CodeLikeACarpenterToolingCommon"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 5`] = `"CodeLikeACarpenterToolingCommon"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 6`] = `"CodeLikeACarpenterFoundationIntermediateRepresentation"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 7`] = `"CodeLikeACarpenterFoundationIntermediateRepresentation"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 8`] = `"CodeLikeACarpenterFoundationIntermediateRepresentation"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 9`] = `"CodeLikeACarpenterFoundationIntermediateRepresentation"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 10`] = `"CodeLikeACarpenterFoundationIntermediateRepresentation"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 11`] = `"AwsOtel"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 12`] = `"AwsOtel"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 13`] = `"AwsOtel"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 14`] = `"AwsOtel"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 15`] = `"AwsOtel"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 16`] = `"CheckRunReporter"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 17`] = `"CheckRunReporter"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 18`] = `"CheckRunReporter"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 19`] = `"CheckRunReporter"`;

exports[`formatStackName() converts a project name to a stack name in ci in prod 20`] = `"CheckRunReporter"`;
